### ISSUE TITLE
bottom of script 3 fails when not run concurrently

### ISSUE THREAD

**Issue by tghoward** [HeritageNetwork/Regional_SDM#36]
If you run script 3 in debugging mode, info at the bottom (such as start_time) doesn't exist?

**Comment by dnbucklin** [comment]
@tghoward @ChristopherTracey This info is generated within `run_SDM` prior to step 3, and if you're manually stepping through from script 2 to script 3, it still wouldn't exist. So if you're debugging script 3, you'll want to at least have started running that script within a `run_SDM` call; that will make sure that metadata information was generated. 

If/when if fails during that step (or any step), then use the lines at the bottom of `user_run_SDM` to refresh/reload arguments from the most recent (failed) run. Then you can step through from the top of any script.

```
# TESTING / DEBUGGING ONLY
library(here)
rm(list=ls())

# for scripts 1-3, run just the following 3 lines
model_species <- "micrmont"
load(here("_data","species",model_species,"runSDM_paths.Rdata"))
for(i in 1:length(fn_args)) assign(names(fn_args)[i], fn_args[[i]])
```

**Comment by tghoward** [comment]
@dnbucklin  so are you saying that no matter where I want to start I need to make a run_SDM call? I just tried that, calling script 3 and I get this:
```
> run_SDM(
+   begin_step = "3",
+   model_species = model_species, # species code in DB; new folder to create in loc_model if not existing
+   loc_scripts = loc_scripts, 
+   nm_presFile = nm_presFile,
+   nm_db_file = nm_db_file, 
+   loc_model = loc_model,
+   loc_envVars = loc_envVars,
+   nm_bkgPts = nm_bkgPts,
+   nm_refBoundaries = nm_refBoundaries, # background grey refernce lines in map
+   nm_studyAreaExtent = nm_studyAreaExtent, # outline black boundary line for study area in map
+   model_comments = model_comments,
+   metaData_comments = metaData_comments,
+   modeller = modeller,
+   add_vars = add_vars,
+   remove_vars = remove_vars,
+   prompt = prompt
+ )
Error in run_SDM(begin_step = "3", model_species = model_species, loc_scripts = loc_scripts,  : 
  Need to begin on step 1 or 2 if adding or removing variables.
> 
```
??

**Comment by tghoward** [comment]
If I send only the lines you pasted (TESTING ...), then begin step 3, the final lines fail.

**Comment by tghoward** [comment]
Trying in reverse order still errors out:
```
> # TESTING / DEBUGGING ONLY
> library(here)
> rm(list=ls())
> # Use the lines below for debugging (running line by line) for a certain script
> # This loads the variables used in previous model run for the species, 
> # so you need to have executed run_SDM in step 2 first.
> 
> # for scripts 1-3, run just the following 3 lines
> model_species <- "bombferv"
> load(here("_data","species",model_species,"runSDM_paths.Rdata"))
> for(i in 1:length(fn_args)) assign(names(fn_args)[i], fn_args[[i]])
> # set wd and load function
> setwd(loc_scripts)
> source(here("helper", "run_SDM.R"))
> run_SDM(
+   begin_step = "3",
+   model_species = model_species, # species code in DB; new folder to create in loc_model if not existing
+   loc_scripts = loc_scripts, 
+   nm_presFile = nm_presFile,
+   nm_db_file = nm_db_file, 
+   loc_model = loc_model,
+   loc_envVars = loc_envVars,
+   nm_bkgPts = nm_bkgPts,
+   nm_refBoundaries = nm_refBoundaries, # background grey refernce lines in map
+   nm_studyAreaExtent = nm_studyAreaExtent, # outline black boundary line for study area in map
+   model_comments = model_comments,
+   metaData_comments = metaData_comments,
+   modeller = modeller,
+   add_vars = add_vars,
+   remove_vars = remove_vars,
+   prompt = prompt
+ )
Error in run_SDM(begin_step = "3", model_species = model_species, loc_scripts = loc_scripts,  : 
  Need to begin on step 1 or 2 if adding or removing variables.
> 
```

**Comment by tghoward** [comment]
I suppose it's OK to say that the tail end of step 3 can't be run when debugging?  But it would be nice to be able to have those variables so I can test the entire thing, line by line.

**Comment by dnbucklin** [comment]
@tghoward The rationale was to save all the inputs to the initial `user_run_SDM` call, but also be able to use a simplified `user_run_SDM` call to pick up the run after step 1. So if you want to start at step 2 or later, you can provide just the arguments needed to start a run from that point. It will automatically pull the last-generated "full run" specifications for that species for all the missing arguments. So to start at step 3 you could do:

```
library(here)
rm(list=ls())

# set project folder and species code for this run
model_species <- "bombferv"
loc_model <- here("_data", "species")
# this is to pull the coded file name for the last prepared presence data
nm_presFile <- gsub("_att.shp","",max(list.files(here("_data","species",model_species,"inputs","model_input"), pattern = "_att.shp")))

# set wd and load function
loc_scripts <- here()
setwd(loc_scripts)
source(here("helper", "run_SDM.R"))

run_SDM(
  begin_step = "3",
  model_species = model_species,
  loc_model = loc_model,
  nm_presFile = nm_presFile
)
```

**Comment by tghoward** [comment]
My bad. I've pulled in your updates and your reference to this issue (https://github.com/HeritageNetwork/Regional_SDM/pull/37#issuecomment-439478992) is correct in that the new changes do take care of it. So just calling the bottom of `user_run_sdm` works fully now. Thanks.   In your very first response to this issue I was hearing you say "it won't exist"   but in reality, it does now exist?

**Comment by tghoward** [comment]
@dnbucklin  it looks like this still fails. try this:
1. run lines 1:80 of `0_user_run_SDM.R`  STOP after running script # 1
2. run the `testing / debugging` lines (86:96), then source script # 2
3. run the `testing / debugging` lines (86:96), then source script # 3

for me, I then get this:

`Error in eval(ei, envir) : object 'modelrun_meta_data' not found`

Am I still doing it wrong?
T

**Comment by dnbucklin** [comment]
You're doing it right; I'm going to move the object. Will PR the change shortly.

**Comment by dnbucklin** [comment]
This should be addressed on all branches as of #97

**Comment by tghoward** [comment]
Worked for me. Thanks!
